src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -367,15 +369,21 @@ async def _handle_unary_unary_rpc(object method_handler,
         loop,
     )
 
-    # Finishes the application handler
-    await _finish_handler_with_unary_response(
-        rpc_state,
-        method_handler.unary_unary,
-        request_message,
-        servicer_context,
-        method_handler.response_serializer,
-        loop
-    )
+    try:
+        # Finishes the application handler
+        await _finish_handler_with_unary_response(
+            rpc_state,
+            method_handler.unary_unary,
+            request_message,
+            servicer_context,
+            method_handler.response_serializer,
+            loop
+        )
+    finally:
+        # Executes done callbacks
+        if rpc_state.done_callbacks:
+            for callback in rpc_state.done_callbacks:
+                callback(servicer_context)
```

Comment:
Im a bit lost with the `servicer_context` parameter, If I understood correctly from the discussion that you had with @gnossen main use case is a cancellation, so the RPC being canceled proactively by the client and by giving a chance to the handler to know when it was canceled by the execution of a callback.

If so the parameter should not be the `RPCState`? why the `ServicerContextz`? does this parameter allows you to know In what terminal state the RPC finished? for example the status code.

Reply:
`RPCState` is a private class for servers, and the client should only see `ServicerContext`.

I think the right path to go should be porting an `AioRpcError` like instance to the done callback.

Reply:
Could we make the `RPCState` attribute of the `ServicerContext` accessible in read-only? So the user if adds a callback for knowing when a request has been canceled for doing whatever would need to know if the requests were really canceled since the `add_done_callback` would be triggered in all of the situations.

Another solution would be provide a specific interface ONLY for when remote cancellation happens, so in that use case the user won't need to check the terminal status of the RPC since the status is gonna be yes or yes the cancellation one.

BTW, personal taste. Other patterns like logging IMO should be implemented through interceptors/middleware.

Reply:
> Could we make the RPCState attribute of the ServicerContext accessible in read-only?

The better design should expose smaller surface API. In done callback, the application can't alter the state of the RPC, so passing the powerful `ServicerContext` object might not be useful. On the other hand, if we allow users to access those value while RPC still running, they might misbehave. So, I hope we can come up with something clean that doesn't require us to introduce a new class.

@pfreixes There is some overlapping between done_callback and interceptors. Since you have more hands-on experience on interceptors, do you think we should provide the done callbacks at all? If the server interceptors are complete, then all discussed scenarios can be fulfilled.

Reply:
IMO the only point where this could supersede the interceptors would be for cleaning up any task related to the RPC, but for that, you would have always the chance of using the `asyncio.CancelledError` exception. So, should the user do something like that: 

```python
async def handler(request, service_context):
    try:
        make_something(...)
    except asyncio.CancelledError:
        clean_stuff_related_to_current_RPC(....)
```

So my expectations would be that in case of a remote cancellation the ongoing RPC in the server would be notified through an `asyncio.CancelledError` exception.

if this is true, having the feeling that the `add_done_callback` wouldn't be necessary.

Reply:
Make sense. I will hold off the `add_done_callback` API. Looking forward to hear updates about the interceptors.